backend/moonstream/routes/streams.py

```diff
@@ -56,54 +62,205 @@
-    except Exception as e:
-        raise HTTPException(status_code=500)
-
-    address_to_subscriptions = {
-        resource.resource_data["address"]: resource.resource_data
-        for resource in user_subscriptions_resources.resources
-    }
-
-    boundaries = data.PageBoundary(
+) -> data.GetEventsResponse:
+    """
+    Gets all events in the client's stream subject to the constraints defined by the following query
+    parameters:
+    - q: Query string which filters over subscriptions
+    - start_time, end_time, include_start, include_end: These define the window of time from which
+    we want to retrieve events.
+
+    All times must be given as seconds since the Unix epoch.
+    """
+
+    stream_boundary = data.StreamBoundary(
         start_time=start_time,
         end_time=end_time,
-        next_event_time=0,
-        previous_event_time=0,
         include_start=include_start,
         include_end=include_end,
     )
 
-    if address_to_subscriptions:
-        response = await actions.get_transaction_in_blocks(
-            db_session=db_session,
-            query=q,
-            user_subscriptions_resources_by_address=address_to_subscriptions,
-            boundaries=boundaries,
-        )
-
-        return response
-    else:
-        return data.EthereumTransactionResponse(stream=[], boundaries=boundaries)
+    user_subscriptions = get_user_subscriptions(request.state.token)
+    query = stream_queries.StreamQuery(
+        subscription_types=[subtype for subtype in event_providers], subscriptions=[]
+    )
+    if q.strip() != "":
+        query = stream_queries.parse_query_string(q)
+
+    _, events = get_events(
+        db_session,
+        bc,
+        MOONSTREAM_DATA_JOURNAL_ID,
+        MOONSTREAM_ADMIN_ACCESS_TOKEN,
+        stream_boundary,
+        query,
+        user_subscriptions,
+        result_timeout=10.0,
+        raise_on_error=True,
+    )
+    response = data.GetEventsResponse(stream_boundary=stream_boundary, events=events)
+    return response
+
+
+@app.get("/latest", tags=["streams"])
+async def latest_events_handler(
+    request: Request, q=Query(""), db_session: Session = Depends(db.yield_db_session)
+) -> List[data.Event]:
+    """
+    Gets the latest events in the client's stream subject to the constraints defined by the following query
+    parameters:
+    - q: Query string which filters over subscriptions
+
+    All times must be given as seconds since the Unix epoch.
+    """
+
+    user_subscriptions = get_user_subscriptions(request.state.token)
+    query = stream_queries.StreamQuery(
+        subscription_types=[subtype for subtype in event_providers], subscriptions=[]
+    )
+    if q.strip() != "":
+        query = stream_queries.parse_query_string(q)
+
+    events = latest_events(
+        db_session,
+        bc,
+        MOONSTREAM_DATA_JOURNAL_ID,
+        MOONSTREAM_ADMIN_ACCESS_TOKEN,
+        query,
+        1,

```

Comment:
Still not sure must we limit ability of `num_events` parameter. right now it return one last element.

Reply:
It returns one element per provider, and handler decides how to deal with that.